### PR TITLE
Add default constructor to Entity class

### PR DESCRIPTION
**Pull request by evie-lau**
## What's changed?
New recipe for OpenJPA -> EclipseLink migration
If class is tagged with `@Entity` or `@MappedSuperclass`, add default no-arg constructor if missing.

## What's your motivation?
https://wiki.eclipse.org/EclipseLink/Examples/JPA/Migration/OpenJPA/Mappings#No-arg_Constructor

### Checklist
- [ ] I've added unit tests to cover both positive and negative cases
- [ ] I've read and applied the [recipe conventions and best practices](https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices)
- [ ] I've used the IntelliJ IDEA auto-formatter on affected files

### CODE REVIEW

File: src/test/java/org/openrewrite/java/migrate/javaee/AddDefaultConstructorToEntityClassTest.java
Comment (timtebeek) [inline src/test/java/org/openrewrite/java/migrate/javaee/AddDefaultConstructorToEntityClassTest.java:179]:
```suggestion
              """,
              """
```

File: src/main/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClass.java
Comment (evie-lau) [inline src/main/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClass.java:82]: This should be generating the correct `public <classname>(){}` constructor, and it did at one point.
But it suddenly started making `public Template <classname>(){}`, and I hadn't made any changes to the recipe code.

Comment (evie-lau) [reply src/main/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClass.java:82]: For reference, 
```
final JavaTemplate defaultConstructorTemplate = JavaTemplate.builder("public #{}(){}").build();
```

Comment (evie-lau) [reply src/main/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClass.java]: Fixed by using `.contextSensitive()` on the JavaTemplate builder

File: src/main/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClass.java
Comment (evie-lau) [reply src/main/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClass.java:24]: @timtebeek Is it actually preferred to use the wildcard import?

Comment (timtebeek) [reply src/main/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClass.java:23]: We try to follow the IntelliJ defaults for formatting and imports, which use a wildcard import for five or more imports from the same class I believe. That way folks can use their short cuts to organize imports and format without giving that too much thought. The automation is here to be consistent going forward, in what I hope is a frictionless experience for the most part.

**Review by timtebeek — APPROVED** [review]
Great work once again! Good to see you worked out the issue with context sensitivity when injecting a constructor. Indeed that'll be down to how we generate and extract the code snippet from JavaTemplate. If you're ever curious what gets generated you can add this little snippet to the code to get the generated template.
image

```java
class __P__ {
    static native <T> T p();
    static native <T> T[] arrp();
    static native boolean booleanp();
    static native byte bytep();
    static native char charp();
    static native double doublep();
    static native int intp();
    static native long longp();
    static native short shortp();
    static native float floatp();
}

class __M__ {
    static native Object any(Object o);

    static native Object any(java.util.function.Predicate<Boolean> o);

    static native <T> Object anyT();
}

public class MissingNoArgConstructorEntity {
    private int id;

    public MissingNoArgConstructorEntity(int id) {
    }

    /*__TEMPLATE__*/
    public MissingNoArgConstructorEntity() {
    }/*__TEMPLATE_STOP__*/
}
```
For constructors specifically we need that context sensitivity, as otherwise the generated template does not compile.
image


Hope that helps you on your next one!

**Comment by timtebeek** [comment]
Once again we'll wait for @cjobinabo to merge when convenient there! :)

File: src/main/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClass.java
Comment (cjobinabo) [inline src/main/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClass.java:47]: I didn't see any mentioned of the `@MappedSuperclass` annotation in the OpenJPA to EclipseLink JPA Migration guide. Did you find additional information that suggested that we should be treating `@MappedSuperclass` the same as `@Entity`?

File: src/test/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClassTest.java
Comment (cjobinabo) [inline src/test/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClassTest.java:27]: We should include a sub-class example where an update is expected here as well.

File: src/main/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClass.java
Comment (evie-lau) [reply src/main/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClass.java:47]: Our rule in WAMT says:
> This rule scans for entity classes with the @Entity or @MappedSuperclass annotations. Entity classes with constructors that do not also have a no-argument constructor are flagged. The quick fix adds an empty no-argument constructor to the class.

Should that be taken to mean that `@MappedSuperclass` still isn't modified?

File: src/test/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClassTest.java
Comment (evie-lau) [reply src/test/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClassTest.java:27]: Do you mean a subclass of an `@Entity` class? and should the subclass be annotated with `@Entity` as well?

File: src/main/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClass.java
Comment (evie-lau) [reply src/main/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClass.java:47]: After we discussed, seems we should add the default constructor to a MappedSuperclass

File: src/main/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClass.java
Comment (evie-lau) [inline src/main/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClass.java:75]: Noticing intermittent discrepencies on the insertion order for `addMethodDeclaration` using this comparator, which causes intermittent unit test failures. 

Maybe should just pick a consistent insertion point for tests? Even if the resulting code might not look as organized...

**Comment by evie-lau** [comment]
Reverted recent change because `@Entity` annotation can never be used on an inner class

File: src/main/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClass.java
Comment (evie-lau) [reply src/main/java/org/openrewrite/java/migrate/javax/AddDefaultConstructorToEntityClass.java]: Switched to using `firstStatement` to guarantee order for tests.

**Pull request merged**
